### PR TITLE
Swift throttle redux

### PR DESCRIPTION
**Pull request by adly-holler**
Here's another go

### CODE REVIEW

**Comment by adly-holler** [comment]
Oh gotta use predicate, hold up

File: ReactiveCocoa/Swift/Signal.swift [hunk 1]
```diff
@@ -569,6 +569,88 @@ public func takeWhile<T, E>(predicate: T -> Bool)(signal: Signal<T, E>) -> Signa
...
+public func throttle<T, E>(interval: NSTimeInterval, onScheduler scheduler: DateSchedulerType, predicate: T -> Bool = { _ in true })(signal: Signal<T, E>) -> Signal<T, E> {
+	precondition(interval >= 0)
+
+	return Signal { observer in
+		let state = Atomic(ThrottleState<T>())
+		let flush = {
+			state.modify { state in
+				if let value = state.pendingValue {
+					let now = scheduler.currentDate
+					sendNext(observer, value)
```
Comment (neilpa) [inline ReactiveCocoa/Swift/Signal.swift]: Forwarding events may be dangerous while holding the lock. Should try to hoist this outside of the modify block

Comment (neilpa) [inline ReactiveCocoa/Swift/Signal.swift]: @jspahrsummers had similar [feedback](https://github.com/ReactiveCocoa/ReactiveCocoa/pull/1714#discussion-diff-24039204) on another PR.

**Comment by adly-holler** [comment]
You're absolutely right, and there's no need to. On it.

File: ReactiveCocoa/Swift/Signal.swift [hunk 2]
```diff
@@ -569,6 +569,88 @@ public func takeWhile<T, E>(predicate: T -> Bool)(signal: Signal<T, E>) -> Signa
...
+							flush()
+						}
+					}
+					compositeDisposable.addDisposable(disposableFromScheduler)
+					newState.scheduleDisposable = ScopedDisposable(compositeDisposable)
+					return newState
+				}
+			} else {
+				state.modify { (var state) in
+					sendNext(observer, value)
```
Comment (neilpa) [inline ReactiveCocoa/Swift/Signal.swift]: Same here about the locking

File: ReactiveCocoa/Swift/Signal.swift [hunk 3]
```diff
@@ -569,6 +569,88 @@ public func takeWhile<T, E>(predicate: T -> Bool)(signal: Signal<T, E>) -> Signa
...
+					if prev == nil || now.timeIntervalSinceDate(prev!) >= interval {
+						scheduleDate = now
+					} else {
+						scheduleDate = prev!.dateByAddingTimeInterval(interval)
+					}
+					
+					let compositeDisposable = CompositeDisposable()
+					let disposableFromScheduler = scheduler.scheduleAfter(scheduleDate) {
+						if !compositeDisposable.disposed {
+							flush()
```
Comment (neilpa) [inline ReactiveCocoa/Swift/Signal.swift]: Same here about the locking

**Comment by jspahrsummers** [comment]
Er, why the use of the predicate? That doesn't match the signature of [the `throttle` that I referenced](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/swift-before-signal-producers/ReactiveCocoa/Swift/HotSignal.swift#L404-L433).

**Comment by adly-holler** [comment]
It was a really handy additional feature in RAC 2 (say you're implementing an autocomplete field and you want to throttle the text signal only if it wasn't cleared) and since we have defaults in Swift it provides no clutter if it isn't used. Want me to remove it?

**Comment by jspahrsummers** [comment]
One of my main concerns with the Swift API is keeping it simple enough that it's not overwhelming. This was a big problem with RAC 2.x, and it'll only be greater with generics and a new language. :wink:

Since we can always add APIs more easily than we can remove or change them, I'd like to leave out the predicate behavior for now. Thanks for thinking of it, though, @adly-holler!

File: ReactiveCocoa/Swift/Signal.swift [hunk 4]
```diff
@@ -569,6 +569,71 @@ public func takeWhile<T, E>(predicate: T -> Bool)(signal: Signal<T, E>) -> Signa
...
+		}, completed: {
+			flush()
+			sendCompleted(observer)
+		})
+	}
+}
+
+// TODO: Move this inside function and rename when compiler stops segfaulting
+private struct ThrottleState<T> {
+	var previousDate: NSDate? = nil
```
Comment (kastiglione) [inline ReactiveCocoa/Swift/Signal.swift]: This is never assigned to.

**Comment by jspahrsummers** [comment]
Is this ready for re-review? (We usually leave an emoji or comment to indicate handoff, since commits don't notify on their own.)

**Comment by adly-holler** [comment]
Oh! Yeah, a review would be great. ⚡

File: ReactiveCocoa/Swift/Signal.swift (hunk 4 above)
Comment (jspahrsummers) [inline ReactiveCocoa/Swift/Signal.swift]: _(Addressed.)_

File: ReactiveCocoa/Swift/Signal.swift [hunk 5]
```diff
@@ -569,6 +569,72 @@ public func takeWhile<T, E>(predicate: T -> Bool)(signal: Signal<T, E>) -> Signa
...
+	var pendingValue: T? = nil
+	var scheduleDisposable: ScopedDisposable? = nil
```
Comment (jspahrsummers) [inline ReactiveCocoa/Swift/Signal.swift]: Why is this a `ScopedDisposable`? TBH, I'm not even sure how that will work inside a value type.

I think it should just be a normal disposable unless there's a reason we can't implement the scoping ourselves.

File: ReactiveCocoa/Swift/Signal.swift [hunk 6]
```diff
@@ -569,6 +569,72 @@ public func takeWhile<T, E>(predicate: T -> Bool)(signal: Signal<T, E>) -> Signa
...
+///
+/// If multiple values are received before the interval has elapsed, the
+/// latest value is the one that will be passed on.
+public func throttle<T, E>(interval: NSTimeInterval, onScheduler scheduler: DateSchedulerType)(signal: Signal<T, E>) -> Signal<T, E> {
+	precondition(interval >= 0)
+
+	return Signal { observer in
+		let state = Atomic(ThrottleState<T>())
+		let flush: () -> Void = {
+			let (_, doSend: (() -> Void)?) = state.modify { (var state) in
```
Comment (jspahrsummers) [inline ReactiveCocoa/Swift/Signal.swift]: Could this maybe yield `T?` instead of the closure, to keep the signature simpler and easier to understand?

**Comment by jspahrsummers** [comment]
Actually, scratch those comments. I'm gonna branch off this PR and make some changes directly, because I've spotted some larger architectural stuff that needs to be reworked.

Since it'll be easier for me to make the changes and then explain them, I'll open a new PR and CC everyone there.

Sorry about the trouble!

**Comment by adly-holler** [comment]
Yo sorry for taking up your time reviewing. I'm still learning this functional-valuetype-Swifty stuff – next PR will be better.

**Comment by jspahrsummers** [comment]
@adly-holler No, it's okay! Sorry, that sounded way more negative than I intended. There's a lot going on here, and most of it is non-obvious (in part because we have zero high-level documentation right now).

I'm going to try to thoroughly explain the changes I make, so hopefully that clarifies any reworking.

**Pull request closed without merging**